## Programavimo įvado pratybos

Prisijungimas ir užduotys:

- Python komandoms prisiminti: ``https://www.w3schools.com/python/``
- Komentarai: #
- Naudingos PowerShell/Terminal komandos:

    - `ls` - pažiūrėti katalogo turinį

    - `md` - sukurti naują katalogą

    - `cd` - pereiti į katalogą / iš katalogo

    - `ni p1.py` - sukuria naują bylą vardu `p1.py`

    - `python p1.py` - kompiliuoja Python programą
   
- Elementarūs matematiniai veiksmai Python:

    - `2*3` - daugyba
    - `2/3` - dalyba
    - `2**3` - kėlimas laipsniu
    - `2**(1/2)` - šaknies traukimas

Visi duomenys Python programavimo kalboje yra objektai. Tai reiškia, kad visi jie yra aprašomi tam tikromis taisyklėmis, kurios nustato ką su tais objektais galima veikti. Ne viską ką galima atlikti su skaičiais, galima atlikti su žodžiais ir atvirkščiai. Pavyzdžiui tam tikro teksto raides galima paversti didžiosiomis `"septyni".upper()`, tačiau to nepavyks padaryti su skaičiumi `7.upper()`.

In [7]:
# Kintamojo tipą galima patikrinti su funkcija type()
x = "vardas"
y = x
z = 7
w = 12.3

z = float(z)
w = str(w)

print (type(x))
print (type(y))
print (type(z))
print (type(w))

<class 'str'>
<class 'str'>
<class 'float'>
<class 'str'>


Pakeiskite aukščiau esančiame pavyzdyje `z` kintamąjį į `float` tipo ir pakeiskite `w` kintamąjį į `string` tipo.

Kintamuosius galime konvertuoti tarp skirtingų tipų pasinaudodami funkcijomis `int()`, `float()` ir `str()`.

In [8]:
skaicius = 123
tekstas = str(skaicius)
skaicius_2 = float(tekstas)

print (type(skaicius))
print (type(tekstas))
print (type(skaicius_2))

<class 'int'>
<class 'str'>
<class 'float'>


Tekstas (string) yra sudėtinis objektas - sudarytas iš atskirų simbolių. Juos galima indeksuoti.

In [13]:
tekstas = "Treciosios programavimo pratybos 2025 metais"

pirma_raide = tekstas[0]

print (pirma_raide)

# Galima pasirinkti ir daugiau nei vieną simbolį panaudojant indeksavimą

ketvirtas_zodis = tekstas[32:37]
data = int (ketvirtas_zodis)

print (data/5)

T
405.0


Modifikuokite viršuje esantį kodą, kad iškirptumėte "2025". Konvertuokite į skaičių, ir padalinkite iš 5.

Sąrašas taip pat yra objektas. Teksto simbolius galime paprastai konvertuoti į sąrašą su funkcija `list()`. Ši funkcija kiekvieną teksto simbolį paverčia atskiru žodyno įrašu. Jeigu norime atskirti ties tam tikrais simboliais, galime naudoti `string.split()` - kur string yra kažkoks tekstas.

In [52]:
tekstas = "To the mind that is still, the whole universe surrenders."

sarasas = list(tekstas)
sarasas2 = tekstas.split(",")   # Tekstą į atskirtas dalis atskiriame ties kableliu
print (sarasas2)

tekstas1 = tekstas.replace("," , "")
zodziai = tekstas1.split (" ")
zodziu_sarasas = list(zodziai)

print (zodziu_sarasas[0:10])

['To the mind that is still', ' the whole universe surrenders.']
['To', 'the', 'mind', 'that', 'is', 'still', 'the', 'whole', 'universe', 'surrenders.']


Modifikuokite viršuje esantį kodą taip, kad sąraše atsidurtų atskiri žodžiai. Tam jums reikės atlikti du veiksmus. Pirmiausia tekste panaikinti kablelį su `string.replace()` metodu, ir tuomet pasinaudojant `string.split()` atskirti ties tarpais (spaces).

Sąrašas taip pat yra sudėtinis objektas (sudarytas iš narių), tad galime juos indeksuoti. Žemiau esančiame kodo pavyzdyje, atskiriame sakinį ties kableliu, vadinasi turime du narius (iki ir po kablelio), ir indeksuodami apsirenkame pirmąją sakinio dalį su indeksu `[0]` - programavime pirmasis narys lygus 0.

In [59]:
tekstas = "To the mind that is still, the whole universe surrenders."

#sarasas = tekstas.split(",")
tekstas1 = tekstas.replace("," , "")
zodziai = tekstas1.split(" ")
zodziu_sarasas = list(zodziai)

print (zodziu_sarasas[0], zodziu_sarasas[1], zodziu_sarasas[2], zodziu_sarasas[6], zodziu_sarasas[7], zodziu_sarasas[8])

To the mind the whole universe


Modifikuokite aukščiau esantį kodą taip, kad pasirinktumėte pirmą, antrą, trečią, septintą, aštuntą ir devintą žodžius.

## A užduotis

#### A1

Jums jau sukurtas kintamasis `letters`, pasinaudodami indeksavimu, išrinkite reikalingas raides ir atspausdinkite savo vardą iš didžiosios raidės.

Užuomina: Indeksavimas vykdomas laužtinių skliaustų pagalba, pvz. `letters[5]` grąžintų raidę `f`.

In [74]:
letters = "abcdefghijklmnopqrstuvwxyz"

print(letters[8].upper(), letters[4], letters[21], letters[0], sep="")


Ieva


#### A2

Sukurkite naują kintamąjį ``auditorija`` ir priskirkite jam savo auditorijos numerį pasinaudodami tik kintamuoju ``digits``, t.y. jei auditorija yra 806, išrinkite atitinkamus skaitmenis.

Užuomina: atkreipkite dėmesį, kad kintamsis `digits` yra skaičius, o indeksavimui reikalingas `string` tipo kintamasis.

In [85]:
digits = 1234567890
skaiciai = str(digits)
skaitmenys = list(skaiciai)

auditorija = str(skaitmenys[7]) + str(skaitmenys[9]) + str(skaitmenys[5])

print (auditorija)

806


#### A3

Jums pateiktas kintamasis `sakinys`. Padalinkite A2 užduotyje sukurtą kintamąjį `auditorija`, iš pateikto sakinio ilgio (simbolių skaičiaus).

```
Auditorijos numeris: [numeris]
Sakinio ilgis: [ilgis]
[numeris] // [ilgis] = [dalyba] (liekana [liekana])
```

Pavyzdžiui, jei auditorijos numeris yra 101, o sakinio ilgis -- 15, tada turite atspausdinti:

```
Auditorijos numeris: 101
Sakinio ilgis: 15
101 // 15 = 6 (liekana 11)

```

In [117]:
sakinys = "Kai programuotojas susiduria su sunkumais ieškodamas klaidų, tai reiškia, kad jis ieško ne ten kur reikia."

auditorija = int (auditorija)

zodziai = list(sakinys)
sakinio_ilgis = len(zodziai)

print("Auditorijos numeris:", auditorija, "\n", "Sakinio ilgis:", sakinio_ilgis, "\n", auditorija, "//", sakinio_ilgis, "=", auditorija//sakinio_ilgis, "(liekana", auditorija%sakinio_ilgis, ")", sep=" ")


Auditorijos numeris: 806 
 Sakinio ilgis: 106 
 806 // 106 = 7 (liekana 64 )


#### A4

Pasinaudokite A3 užduotyje pateiktu kintamuoju `sakinys`. Pakeiskite visas "s" raides skaičiumi "5", ir visas raides į didžiąsias.

In [118]:
naujas_sakinys = sakinys.replace("s", "5").upper()

print(naujas_sakinys)

KAI PROGRAMUOTOJA5 5U5IDURIA 5U 5UNKUMAI5 IEŠKODAMA5 KLAIDŲ, TAI REIŠKIA, KAD JI5 IEŠKO NE TEN KUR REIKIA.


#### A5

Parašykite kodą, kuris paprašytų įvesti žodį, ir atspausdintų "Jūs įvedėte žodį "..." kurio ilgis yra ... .

Užuomina: Pasinaudokite `input()`, `len()` funkcijomis.

In [111]:
zodis = input("Įveskite žodį: ")
ilgis = len(zodis)

print ("Jūs įvedėte žodį ", '"', zodis, '"', " kurio ilgis yra ", ilgis, sep="")



Įveskite žodį: Labas
Jūs įvedėte žodį "Labas" kurio ilgis yra 5


#### A6

Parašykite kodą, kuris paprašytų įvesti betkokį natūralųjį skaičių, tuomet padalinkite duotąjį skaičių iš to skaičiaus veidrodinės versijos (pvz. 457 veidrodinė versija yra 754). Atspausdinkite "Jūsų įvestas skaičius yra ..., padalinus jį iš ... gauname ..."

In [124]:
skaicius = input ("Įveskite norimą natūralūjį skaičių: ")
skaicius = int(skaicius)

skaitmenys = str(skaicius)
#atvirkscias = int[::-1] ?
atvirkscias = int (atvirkscio_skaitmenys)

#print(tekstas[(einu nuo kur):(einu iki kur):(dedu zingsni)])
#print[::-1] - eis nuo galo



print ("Jūsų įvestas skaičius yra: ", skaicius, ", padalinus jį iš ", atvirkscio_skaitmenys, " gauname ", skaicius/atvirkscias, sep="")

Įveskite norimą natūralūjį skaičių: 123
Jūsų įvestas skaičius yra: 123, padalinus jį iš 321 gauname 0.38317757009345793


#### A7

Jums sukurtas kintamasis, tam tikro duomenų failo antraštinė eilutė su informacija.

Parašykite kodą, kuris iš šio kintamojo iškerpa visas penkias skaitines vertes ir priskiria jas naujiems kintamiesiems. Pasinaudokite galimybe karpyti (slicing). Skaitines vertes atspausdinkite.

Dažniausiai duomenų failai yra standartizuoti, tad parašę kodą kuris iškerpa tam tikrą informaciją iš vieno failo, tas pats kodas veiks ir su kitais atitinkamais duomenų failais.

In [133]:
eilute = "Radial velocity =   -55.423  dispersion =   0.902  FWHM =  12.129, 912 lines matched out of 5014"

zodziai = eilute.replace(",", "").replace("=", "").split()
#print(zodziai)

Rv = zodziai[2]
Dis = zodziai[4]
FWHM = zodziai[6]
line_no = zodziai[7]
total_line_no = zodziai[12]

# Dis =
# FWHM =
# line_no =
# total_line_no =

print (Rv, Dis, FWHM, line_no, total_line_no)

-55.423 0.902 12.129 912 5014


#### A8

Dar kartą atkreipkite dėmesį į duomenų failo antraštę. Joje pateikta informacija apie linijas (šiam momentui nėra svarbu kas tai yra).

Apskaičiuokite kiek procentų linijų buvo rasta (matched). Rezultatą atspausdinkite 2 skaičių po kablelio tikslumu. Nepamirškite atspausdinti procento ženklo, pvz. 12.30% (jeigu antras skaičius po kablelio lygus 0, jį irgi atspausdinti reikia). Pasinaudokite ankstesnėje užduotyje sukurtais kintamaisiais.

In [137]:
percentage = int (line_no)/int(total_line_no)*100

print(f"Rasta {percentage:.2f}% lnijų")

Rasta 18.19% lnijų


#### A9

Pridėkite prie kintamojo ``eilute`` informaciją, kurią suskaičiavote 6-oje užduotyje. Rezultatą atspausdinkite.

Atspausdintas rezultatas turėtų atrodyti panašiai: `"Radial velocity =   -55.423  dispersion =   0.902  FWHM =  12.129, 912 lines matched out of 5014 (18.19%)"`.

#### A10

Jums pradėtas kurti naujas sąrašas (list). Papildykite jį šios dienos datos elementais. Pvz. 20250925

Tuomet sąrašą išrikiuokite didėjimo tvarka, naudokite metodą `sort()` arba funkciją `sorted()`. (Metodas naudojamas "ant" kintamojo `sarasas.sort()`, funkcijai kintamasis paduodamas `sorted(sarasas)`.

In [ ]:
numbers = [2, 0, 2, 5, ....]


#### A11

Raskite didžiausią ir mažiausią `numbers` sąrašo vertes, apskaičiuokite sąraše esančių skaičių vidurkį.

Užuomina: Pasinaudokite `min()`, `max()`, `sum()` ir `len()` funkcijomis.

#### A12

Paprašykite įvesti betkokį sveikąjį teigiamą skaičių. Pateiktas kodas pateiktą skaičių konvertuoja į simbolių sąrašą. Pataisykite kodą, kad simboliai būtų skaičiai type: int, o ne type: str. A10 užduotyje sukurtą sąrašą `numbers` papildykite naujais skaitmenimis iš sąrašo `numbers_2`. Vėl išrikiuokite sąrašą didėjimo tvarka.

In [ ]:
ivestas_skaicius =                  # Papildykite šią eilutę kodu, kuri paprašytų įvesti sveikąjį skaičių.
numbers_2 = []                      # Sukurtas naujas tuščias sąrašas į kurį dėsime naujas vertes.

for skaitmuo in ivestas_skaicius:   # Imame kiekvieną įvesto skaičiaus skaimenį
  numbers_2.append(skaitmuo)        # Pridedame kiekvieną skaitmenį į naujai sukurtą number_2 sąrašą
print (numbers_2)

# Papildykite kodu kuris kintamąjį numbers papildytų nariais iš numbers_2.


#### A13

Suskaičiuokite kiek narių yra jūsų naujajame sąraše. (`len()`)

Pakeiskite n-tąjį narį verte m (paprašykite įvesti vertes n ir m). (`input()`). m turėtų būti vienaženklis skaičius. Sąrašą vėl išrikiuokite didėjimo tvarka.

Pvz:

"Dabar sąraše yra 15 skaitmenų, kurį skaitmenį norite pakeisti? "

10

"Kokį skaitmenį norite įrašyti į 10 vietą? "

9

## B užduotis

Parašykite primityvų kriptografijos algoritmą:

#### B1

Sukurkite kintamąjį:

```python
letters = "abcdefghijklmnopqrstuvwxyz "
message = "slapti egzamino uzduociu atsakymai"
```

In [ ]:
letters = "abcdefghijklmnopqrstuvwxyz "
message = "slapti egzamino uzduociu atsakymai"

#### B2

Sumaišykite kintamojo ``letters`` raidžių tvarką. Rezultatą užsaugokite naujame kintamajame ``code``.

Užuomina: https://stackoverflow.com/questions/2668312/shuffle-string-in-python

#### B3

Užkoduokite kintamąjį ``message`` pakeisdami raides pagal ``code`` tvarką.

Pavyzdžiui, jei ``code = "ijk..."``, tada raidė ``a`` turi būti pakeista raide ``i``, raidė ``b`` - raide ``j``, raidė ``c`` - raide ``k``, ir t.t.

Rezultatą atspausdinkite. Užuominos:

https://www.w3schools.com/python/python_for_loops.asp

https://stackoverflow.com/questions/2294493/how-to-get-the-position-of-a-character-in-python

#### B4

Pasinaudodami kodu

```python
code = "zxfdbhtja epqucovgwlsykmnir"
```

atkoduokite žinutę

```python
message = "zlb zsrozqzfazsrustzpb zs"
```

Rezultatą atspausdinkite.

In [ ]:
code    = "zxfdbhtja epqucovgwlsykmnir"
letters = "abcdefghijklmnopqrstuvwxyz "
message = "zlb zsrozqzfazsrustzpb zs"

## Sandbox